### PR TITLE
chore: move xhr management to a module and refactor

### PR DESCRIPTION
**Pull request by yohanboniface**
It sort of work for common cases, but edge cases needs work, specifically the login flow.

- [x] better error handling
- [x] 412
- [ ] dependency on UI
- [ ] login required when saving the map/datalayers

### CODE REVIEW

File: umap/static/umap/js/modules/request.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
+// Uses `L._`` from Leaflet.i18n which we cannot import as a module yet
+import { Evented, DomUtil } from '../../vendors/leaflet/leaflet-src.esm.js'
+
+export class HTTPError extends Error {
```
Comment (almet) [inline umap/static/umap/js/modules/request.js:4]: Should all the exception inherit from the same base class, so we can intercept all the ones from this module alltogether?

File: umap/static/umap/js/modules/request.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  constructor(message) {
+    super(message)
+    this.name = this.constructor.name
+  }
+}
+
+export class NokError extends Error {
```
Comment (almet) [inline umap/static/umap/js/modules/request.js:11]: nit: should this be in caps (`NOKError`) ?

Comment (almet) [reply umap/static/umap/js/modules/request.js:11]: Documenting what's the difference between `HTTPError` and `NOKError` could be useful :-)

File: umap/static/umap/js/modules/request.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    } catch (error) {
+      console.error(error)
+      this.fire('dataload', { id: id })
+      throw new HTTPError(error.message)
+    }
+    if (!response.ok) {
+      this.fire('dataload', { id: id })
+      throw new NOKError(response)
+    }
+    // TODO
```
Comment (davidbgk) [inline umap/static/umap/js/modules/request.js:45]: I guess it's done?

File: umap/static/umap/js/modules/request.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+export class NOKError extends RequestError {
+  constructor(response) {
+    super(response.status)
+    this.response = response
+    this.status = response.status
+    this.name = this.constructor.name
+  }
+}
+
+const BaseRequest = Evented.extend({
```
Comment (davidbgk) [inline umap/static/umap/js/modules/request.js:23]: I'm still wondering if it's a good idea to extend the Leaflet logic here, let's talk one last time about that :)

File: umap/static/umap/js/umap.js [hunk 5]
```diff
@@ -1093,64 +1061,62 @@ L.U.Map.include({
...
-                callback: () => {
-                  L.Util.copyToClipboard(data.permissions.anonymous_edit_url)
-                  this.ui.alert({
-                    content: L._('Secret edit link copied to clipboard!'),
-                    level: 'info',
-                  })
-                },
-                callbackContext: this,
+    const uri = this.urls.get('map_save', { map_id: this.options.umap_id })
+    const [data, response, error] = await this.server.post(uri, {}, formData)
```
Comment (almet) [inline umap/static/umap/js/umap.js:1065]: Could it be possible to do the `this.fire('dataloading', e))` here? (I'm trying to see if we could get rid of events at the request level)

File: umap/static/umap/js/umap.js [hunk 6]
```diff
@@ -1093,64 +1061,62 @@ L.U.Map.include({
...
+    if (!error) {
+      let duration = 3000,
```
Comment (almet) [inline umap/static/umap/js/umap.js:1067]: (and here for the  `dataload` maybe?)

**Review by davidbgk — APPROVED** [review]

**Pull request merged**